File: csharp-blazor-webgl/Lib/Math/Vector3.cs
using System.Numerics;

namespace BlazorExperiments.Lib.Math;

public struct Vector3<T> :
    IAdditionOperators<Vector3<T>, Vector3<T>, Vector3<T>>,
    ISubtractionOperators<Vector3<T>, Vector3<T>, Vector3<T>>,
    IMultiplyOperators<Vector3<T>, T, Vector3<T>>,
    IDivisionOperators<Vector3<T>, T, Vector3<T>>,
    IUnaryNegationOperators<Vector3<T>, Vector3<T>>
    where T : INumber<T>
{
    public readonly T X;
    public readonly T Y;
    public readonly T Z;

    public Vector3(T x, T y, T z)
    {
        X = x;
        Y = y;
        Z = z;
    }

    public override string ToString()
    {
        return $"({X}, {Y}, {Z})";
    }

    public static Vector3<T> operator +(Vector3<T> left, Vector3<T> right)
    {
        return new(
            left.X + right.X,
            left.Y + right.Y,
            left.Z + right.Z
        );
    }

    public static Vector3<T> operator -(Vector3<T> left, Vector3<T> right)
    {
        return new(
            left.X - right.X,
            left.Y - right.Y,
            left.Z - right.Z
        );
    }

    public static Vector3<T> operator *(Vector3<T> left, T right)
    {
        return new(
            left.X * right,
            left.Y * right,
            left.Z * right
        );
    }

    public static Vector3<T> operator /(Vector3<T> left, T right)
    {
        return new(
            left.X / right,
            left.Y / right,
            left.Z / right
        );
    }

    public static Vector3<T> operator -(Vector3<T> value)
    {
        return new(-value.X, -value.Y, -value.Z);
    }

    public T MagnitudeSquared => X * X + Y * Y + Z * Z;

    public T DotProduct(Vector3<T> other)
    {
        return X * other.X + Y * other.Y + Z * other.Z;
    }

    public Vector3<T> CrossProduct(Vector3<T> other)
    {
        return new(
            Y * other.Z - Z * other.Y,
            Z * other.X - X * other.Z,
            X * other.Y - Y * other.X
        );
    }
}

public static class Vector3Extensions
{
    public static T GetMagnitude<T>(this Vector3<T> v) where T : INumber<T>, IRootFunctions<T>
    {
        return T.Sqrt(v.MagnitudeSquared);
    }

    public static Vector3<T> Normalized<T>(this Vector3<T> v) where T : INumber<T>, IRootFunctions<T>
    {
        return v / v.GetMagnitude();
    }

    public static Radians<T> AngleBetween<T>(this Vector3<T> v, Vector3<T> other) where T : INumber<T>, IRootFunctions<T>, ITrigonometricFunctions<T>
    {
        /*
        https://stackoverflow.com/a/16544330/9290998
        https://stackoverflow.com/a/67719217/9290998
        x = dot(a, b)
        y = dot(n, cross(a, b))
        angle = atan2(y, x)
        */
        return TrigExtensions.Atan2(
            v.CrossProduct(other).GetMagnitude(),
            v.DotProduct(other)
        );
    }
}

File: swift-c-interop-sdl3-wgpu/Sources/Experiment/math-and-graphics-types/Vector3.swift
struct Vector3<T> {
    let x: T
    let y: T
    let z: T
}

extension Vector3 where T: Mathable {
    static func + (left: Self, right: Self) -> Self {
        Self(
            x: left.x + right.x,
            y: left.y + right.y,
            z: left.z + right.z
        )
    }

    static func - (left: Self, right: Self) -> Self {
        Self(
            x: left.x - right.x,
            y: left.y - right.y,
            z: left.z - right.z
        )
    }

    static func * (left: Self, right: T) -> Self {
        Self(
            x: left.x * right,
            y: left.y * right,
            z: left.z * right
        )
    }

    static func * (left: T, right: Self) -> Self {
        Self(
            x: left * right.x,
            y: left * right.y,
            z: left * right.z
        )
    }

    static func / (left: Self, right: T) -> Self {
        Self(
            x: left.x / right,
            y: left.y / right,
            z: left.z / right
        )
    }

    static prefix func + (unary: Self) -> Self {
        unary
    }

    static prefix func - (unary: Self) -> Self {
        Self(
            x: -unary.x,
            y: -unary.y,
            z: -unary.z
        )
    }

    static func dot(_ left: Self, _ right: Self) -> T {
        left.x * right.x + left.y * right.y + left.z * right.z
    }

    static func cross(_ left: Self, _ right: Self) -> Self {
        Self(
            x: left.y * right.z - left.z * right.y,
            y: left.z * right.x - left.x * right.z,
            z: left.x * right.y - left.y * right.x
        )
    }

    var magnitudeSquared: T { x * x + y * y + z * z }
}

extension Vector3 where T: Mathable & Sqrt {
    var magnitude: T { magnitudeSquared.sqrt }

    var normalized: Self { self / magnitude }
}

// TODO also do this for 2d and 4d vectors
extension Vector3 where T: Mathable & Sqrt & Trigonometry {
    static func angleBetween(_ a: Vector3<T>, _ b: Vector3<T>) -> Radians<T> {
        /*
        https://stackoverflow.com/a/16544330/9290998
        https://stackoverflow.com/a/67719217/9290998
        x = dot(a, b)
        y = dot(n, cross(a, b))
        angle = atan2(y, x)
        */
        return T.atan2(
            y: cross(a, b).magnitude,
            x: dot(a, b)
        )
    }
}
